### PR TITLE
#2873: Check for presence of execinfo.h before enabling its use for stack tracing

### LINKED ISSUE #2873: build on musl libc fails due to missing execinfo.h used for stack tracing
On system using musl libc the build fails with:
```
/usr/bin/x86_64-gentoo-linux-musl-g++ -Dkokkoscore_EXPORTS -Icore/src -I/var/tmp/portage/dev-cpp/kokkos-3.0.00/work/kokkos-3.0.00/core/src -I.  -DNDEBUG -O2 -pipe -march=native -fno-common -falign-functions=32:25:16 -fPIC   -fopenmp -MD -MT core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_Stacktrace.cpp.o -MF core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_Stacktrace.cpp.o.d -o core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_Stacktrace.cpp.o -c /var/tmp/portage/dev-cpp/kokkos-3.0.00/work/kokkos-3.0.00/core/src/impl/Kokkos_Stacktrace.cpp
/var/tmp/portage/dev-cpp/kokkos-3.0.00/work/kokkos-3.0.00/core/src/impl/Kokkos_Stacktrace.cpp:6:10: fatal error: execinfo.h: No such file or directory
    6 | #include <execinfo.h>
      |          ^~~~~~~~~~~~
compilation terminated.
```
as musl doesn't have a `execinfo.h`.

See https://bugs.gentoo.org/712748

### PR DESCRIPTION
**Pull request by PhilMiller**
Fixes #2873

### CODE REVIEW

**Comment by PhilMiller** [comment]
retest this please

File: core/src/Kokkos_Macros.hpp [hunk 1]
```diff
@@ -601,7 +601,9 @@ static constexpr bool kokkos_omp_on_host() { return false; }
 #if (defined(KOKKOS_COMPILER_GNU) || defined(KOKKOS_COMPILER_CLANG) ||  \
```
[the remark this answers is not in the document]
Comment (PhilMiller) [reply core/src/Kokkos_Macros.hpp:601]: Yeah, that `#if` condition is basically "which libc are we compiling against?" GNU (and maybe BSD) would be OK, musl, MSVCRT, and Bionic are not.

A direct check for the functionality we want seems much more straightforward.

[the remark this answers is not in the document]
Comment (PhilMiller) [reply core/src/Kokkos_Macros.hpp:601]: Speaking of which, do we have any header availability checks in our CMake? If so, we could just move them straight into the code.

[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/Kokkos_Macros.hpp:601]: [`CHECK_INCLUDE_FILE_CXX`](https://cmake.org/cmake/help/v3.16/module/CheckIncludeFileCXX.html#module:CheckIncludeFileCXX)

**Review by crtrott — APPROVED** [review]

**Review by cz4rs — APPROVED** [review]
Looks good to me.

**Pull request merged**